### PR TITLE
Prevent marking broken connections as verified

### PR DESCRIPTION
**Pull request by composerinteralia**
Related to https://github.com/rails/rails/pull/49802 (we may want both PRs)

Prior to this commit `with_raw_connection` would always mark the connection as `@verified = true` after yielding it. But yielding a connection is not enough to determine that a connection is healthy. Marking a broken connection as verified can prevent further reconnects, leaving us with a broken connection for longer than expected (including into the next request).

For example,

* Connection is broken and not verified
* Quote string succeeds, and marks the broken connection as verified
* Because the connection is verified, querying will not trigger a reconnect
* Query fails and the connection is no longer verified
* Repeat ad infinitum (or until we reach a query with `allow_retry: true`, or one that isn't preceded by a call to quote string)

This commit solves the problem by only marking the connection as verified if we have definitively exercised that connection. This is safer, but it does put more responsibility on each individual adapter. If a call to `verified!` is missing things should still work OK, but we'll end up pinging the connection more to see if it is active.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

cc @matthewd

### CODE REVIEW

**Comment by byroot** [comment]
Hum, intuitively I would instead make this a parameter of `with_raw_connection`. So it keeps verifying by default, but you can just pass `verify: false` when doing things like quoting.

**Comment by composerinteralia** [comment]
> keeps verifying by default

Sure, I can do that way if it's preferred. My thinking in NOT verifying by default is that the consequences of missing a caller are slightly better (pinging the connection too much vs. verifying a broken connection).

**Comment by byroot** [comment]
Good point. What about making it a mandatory parameter?

**Comment by composerinteralia** [comment]
I like that idea. Would that make the change more difficult to backport to 7.1, since it may break other adapters?

**Comment by byroot** [comment]
> may break other adapters?

There isn't many of those. We can check the Oracle and MSSQL ones see if they'd be broken.

But we can also make it default to `verify: true` on 7.1 to preserve that compat and make it required in `main`, it doesn't make the backport that much harder.

**Comment by matthewd** [comment]
Hmmm... as verification is essentially an optimization, I don't love the prominence it gets from being a required parameter. On average each adapter is only going to have a couple of methods that actually talk to the remote server, so I think it seems clearer to leave it to those callers to affirmatively declare "I just heard from the server, it's definitely there" rather than having them pre-commit to "I promise that in this upcoming block, I _will_ talk to the server and wait for its response".

While we don't have a ton of call sites of either variety to sample, I think the current diff bears out my expectation that if you're adding some new off-the-path reason to need to grab a raw connection, whatever you're doing _probably_ doesn't warrant inferred-verification. Only SQLite seems to show a preponderance of `verify: true`, and that's a special case because it doesn't really have a connection to verify in the first place.

Finally, it's a super-tiny point, but in-the-moment verification could also play slightly nicer with nested blocks (`with(allow_retry: true) { query; with(allow_retry: false) { .. } }`), though realistically they should almost certainly be avoided for other reasons.

---

Not unrelated, `verify` as a parameter name is not sufficiently specific -- this is "when my block completes, you may infer the connection is now verified", which is separate and arguably less intuitive than "verify the connection before you enter the block" [which is not directly configurable, but is IMO more likely the behaviour a caller might think of first, and could easily lead to "yes please verification sounds good", reintroducing this problem].

I was going to suggest a `verified!`-in-main, `verify: true`-for-backport approach, but then I realised that I think `verified!` would actually also make for the better backport strategy: third-party-adapter-using upgraders will start issuing needless verification queries, but will be immediately safe, and then the adapters can maintain compatibility by dropping `verified! if respond_to?(:verified!)` into their execute method(s). Backport considerations should not drive our `main` design decisions, but, independently, I think that ends up more manageable than a new kwarg that you [as an adapter] are pretty much obliged to sometimes provide.

**Comment by byroot** [comment]
> so I think it seems clearer to leave it to those callers to affirmatively declare "I just heard from the server, it's definitely there" rather than having them pre-commit to "I promise that in this upcoming block, I will talk to the server and wait for its response".

The reason I prefer the parameter is precisely because it's an optimization. Marking the connection is verified is likely to be an "unknown unknown" concern for future contributors, or people extending an adapter. By making it a required parameter we turn it into a "known unknown".

We could of course consider that since it's an optimization it's no big deal if someone forget to verify where it could be done, but I'd hate having to debug this.

**Comment by matthewd** [comment]
Yeah, that's a fair point. The overly-ambitious current fully-automatic behaviour is itself a result of me trying to avoid the unknown-unknown scenario.

We can assert that our basic "run a query" methods leave the adapter verified, so that will prevent it falling off entirely, or being completely unnoticed for a brand new adapter. Beyond that, I think I'm happy with the idea that someone might miss a possible verification, and so a connection that does that retryable technically-verifying operation followed by a non-retryable action will insert a `SELECT 1` in between when it could've been skipped -- it'll be failing safe, but should have less risk of someone (even someone who thinks they understand this stuff quite well 👋🏻😳) believing the upcoming block is a guaranteed verify when that's not quite true. (Similarly for future maintenance, `verified!` on the line directly after `conn.exec_params` is perhaps slightly more noticeable than one of the options on a call that could be a few lines / indentation levels above.)

**Comment by matthewd** [comment]
> which would interfere with the still-yielded connection in the outer block

That was indeed the "other reasons" I was picturing. (Thanks for documenting, past me!)

I'm torn because on one hand, if we want to say "nesting bad" we could trivially check and enforce that, but it also feels potentially-annoying and restrictive, especially for a moderately-internal method, where we would generally tend more toward "you get to keep both pieces". 

I believe nesting is safe, as long as the inner call disables retries (or, technically, [maybe] if the outer call doesn't touch its yielded `conn` after the inner call occurs... but then, why is the block still open?) -- so maybe it's fine to leave it as possible-but-discouraged?

I didn't do a full sweep, but I just got as far as noting `with_multi_statements`, which doesn't yield to user code, but does wrap calls that use user-facing methods (thus ultimately nesting `with_raw_connection`)... it holds its raw connection with reason (ensuring the state change doesn't escape), but it still seems pretty inadvisable.

Avoiding a reconnection while an outer block is outstanding doesn't seem horribly complicated, so maybe that's the best way to go -- avoid where feasible just because it could get confusing, and/or might prevent a recovery that might otherwise be possible, but also make it properly safe when it's judged necessary.

**Comment by composerinteralia** [comment]
I probably should have mentioned that the specific case I uncovered during my sweep was
https://github.com/rails/rails/blob/344e59e1d1ca92703aef3947ef1e0e28abb6709c/activerecord/lib/active_record/connection_adapters/sqlite3/database_statements.rb#L69-L78. I was imagining we'd lose `read_uncommitted = true` if the inner block reconnected, but maybe that's not a realistic concern with sqlite anyway.

**Comment by composerinteralia** [comment]
Sorry, got a little off topic. Dealing with nested `with_raw_connection` reconnects is a separate issue, so we can explore that more in a separate PR.

Anything you'd like me to change in this PR?

File: activerecord/lib/active_record/connection_adapters/abstract_adapter.rb
Comment (matthewd) [inline activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:816]: As the PR needs a rebase anyway: I think we can push this down to `private`

**Comment by nvasilevski** [comment]
Hey @composerinteralia, just wanted to continue our conversation here https://github.com/trilogy-libraries/trilogy/pull/134 without creating noise in the trilogy PR. 

So I've been following https://github.com/rails/rails/issues/49781 and all related PRs as the problem does sound extremely similar to what we are seeing in production but I was sceptical because at first it was scoped to postgresql and wasn't easily reproducible on mysql for both mysql and trilogy along with us now knowing why the issue is only happening on `trilogy` as the adapters for mysql2 and trilogy are almost the same.

However, I tried this commit and indeed it fixed our problem which means we had our connections marked as `@verified` by most likely `quote_string`. The only question that is left why the issue doesn't happen when using `mysql2` adapter and I think I have an answer: `mysql2` checks the connection state before performing the `escape` and raises an exception which is being caught by Rails and leads to the connection to be correctly marked as `@verified=false` 
https://github.com/brianmario/mysql2/blob/79f78f940685396f1b5f30ec502544bb7e3ba9cf/ext/mysql2/client.c#L908
while trilogy just performs the escaping if I'm following the code correctly
https://github.com/trilogy-libraries/trilogy/blob/e577853a1f40a681cd7fc91c3c4e97e69e31762c/contrib/ruby/ext/trilogy-ruby/cext.c#L951
https://github.com/trilogy-libraries/trilogy/blob/e577853a1f40a681cd7fc91c3c4e97e69e31762c/src/client.c#L667

Which means if this assumption is correct the test you added here should not fail when executed on `mysql2` even without the fix while on `trilogy` it will fail but pass after the fix. (I'm going to poke around it a little more)

But basically I wanted to say that this PR is a solution to a real problem (at least when using `trilogy`) and thank you for making the change!

On a side note, does anybody think that `trilogy` should also perform the connection check on `escape` as `mysql2` does using `REQUIRE_CONNECTED`? I think this would have been fixed the problem by implication even though it shouldn't be a substitution for this PR.

**Comment by nvasilevski** [comment]
>the test you added here should not fail when executed on mysql2 even without the fix while on trilogy it will fail but pass after the fix.

I was slightly wrong and while the test your added fails on `main` for both trilogy and mysql2 here is the test which will pass on `main` on `mysql2` but fail on `trilogy` essentially revealing the difference and reproducing the exact problem we are seeing in production:

```ruby
      test "recovers from a disconnect when going through `quote_string`" do
        Post.first # Connection verified (and prepared statement pool populated if enabled)
        remote_disconnect @connection

        assert_raises(ActiveRecord::ConnectionFailed) do
          Post.where(["title=?", "a"]).to_a
        end

        # recovers
        Post.where(["title=?", "a"]).to_a
      end
```
`Post.where(["title=?", "a"]).to_a` can be substituted with an explicit call to `connection.quote_string` + any following query but this example is being a little closer to a real one.
On trilogy queries that go through `quote_string` will never recover since `connection#escape` doesn't raise and doesn't follow the `allow_retry: true` branch unlike `mysql2`

**Comment by matthewd** [comment]
> On a side note, does anybody think that trilogy should also perform the connection check on escape as mysql2 does using REQUIRE_CONNECTED?

Personally: not really. https://github.com/csfrancis/mysql2/pull/1 shows some opinion that that behaviour may be undesirable in mysql2 -- and I've previously noted in https://github.com/rails/rails/pull/44576#discussion_r822476418 that I think pg (libpq?) is being needlessly particular.

It does make those adapters less susceptible to this particular bug, but I think that's more happenstance than intentional design.

**Review by rafaelfranca — APPROVED** [review]
@matthewd any blockers to merge and backport this?

**Pull request merged**
